crates/utils/re_video/src/stable_index_deque.rs

```diff
@@ -0,0 +1,322 @@
+/// This is useful to use stable indices for addressing in a growing/shrinking deque
+/// without having to resort to a more complex map datastructure.
+///
+/// To illustrate:
+/// ```
+/// // Indices for deque are unstable:
+/// # use std::collections::VecDeque;
+/// let mut v = (0..2).collect::<VecDeque<i32>>();
+/// assert_eq!(v.get(1), Some(&1));
+/// v.pop_front();
+/// assert_eq!(v.get(1), None);
+///
+/// // Indices for `StableIndexDeque` are stable:
+/// # use re_video::StableIndexDeque;
+/// let mut v = (0..2).collect::<StableIndexDeque<i32>>();
+/// assert_eq!(v.get(1), Some(&1));
+/// v.pop_front();
+/// assert_eq!(v.get(1), Some(&1));
+/// ```
+#[derive(Default, Clone, Debug)]
+pub struct StableIndexDeque<T> {
+    vec: VecDeque<T>,
+    index_offset: usize,
+}
+
+impl<T> StableIndexDeque<T> {
+    #[inline]
+    pub fn new() -> Self {
+        Self {
+            vec: VecDeque::new(),
+            index_offset: 0,
+        }
+    }
+
+    /// Creates a new deque from an iterator and an index offset.
+    ///
+    /// ```
+    /// # use re_video::StableIndexDeque;
+    /// let v = StableIndexDeque::from_iter_with_offset(0..2, 1);
+    /// assert_eq!(v.get(0), None);
+    /// assert_eq!(v.get(1), Some(&0));
+    /// assert_eq!(v.get(2), Some(&1));
+    /// ```
+    pub fn from_iter_with_offset(iter: impl IntoIterator<Item = T>, index_offset: usize) -> Self {
+        Self {
+            vec: VecDeque::from_iter(iter),
+            index_offset,
+        }
+    }
+
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            vec: VecDeque::with_capacity(capacity),
+            index_offset: 0,
+        }
+    }
+
+    /// See [`VecDeque::push_back`].
+    #[inline]
+    pub fn push_back(&mut self, value: T) {
+        self.vec.push_back(value);
+    }
+
+    /// Unlike with [`VecDeque::pop_front`], indices into the deque stay the same.
+    ///
+    /// ```
+    /// # use re_video::StableIndexDeque;
+    /// let mut v = (0..2).collect::<StableIndexDeque<i32>>();
+    /// assert_eq!(v.get(0), Some(&0));
+    /// assert_eq!(v.get(1), Some(&1));
+    /// v.pop_front();
+    /// assert_eq!(v.get(0), None);
+    /// assert_eq!(v.get(1), Some(&1));
+    /// ```
+    #[inline]
+    pub fn pop_front(&mut self) -> Option<T> {
+        self.vec.pop_front().inspect(|_| self.index_offset += 1)
+    }
+
+    /// See [`VecDeque::pop_back`].
+    pub fn pop_back(&mut self) -> Option<T> {
+        self.vec.pop_back()
+    }
+
+    /// See [`VecDeque::extend`].
+    #[inline]
+    pub fn extend(&mut self, values: impl IntoIterator<Item = T>) {
+        self.vec.extend(values);
+    }
+
+    /// See [`VecDeque::iter`].
+    #[inline]
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        self.vec.iter()
+    }
+
+    /// See [`VecDeque::iter_mut`].
+    #[inline]
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut T> {
+        self.vec.iter_mut()
+    }
```

Comment:
We should warn off users from doing `.iter().enumerate()` and maybe provide a helper (`.iter_indexed()` or similar) that takes the offset into account

Reply:
oh yes this would actually be really useful in a bunch of places!